flink-python/dev/dev-requirements.txt

```diff
@@ -15,20 +15,20 @@
 pip>=20.3
 setuptools>=18.0
 wheel
-apache-beam==2.43.0
-cython==0.29.24
+apache-beam>=2.43.0,<2.49.0
+cython>=0.29.24
 py4j==0.10.9.7
 python-dateutil>=2.8.0,<3
-cloudpickle==2.2.0
-avro-python3>=1.8.1,!=1.9.2,<1.10.0
-pandas>=1.3.0,<1.4.0
-pyarrow>=5.0.0,<9.0.0
+cloudpickle>=2.2.0
```

Comment:
Hi @HuangXingBo! Thank you for reviewing.

I tested it out, and, unfortunately, I can't resolve dependencies between Ibis and PyFlink (based on the loosened dependencies in this PR) if I pin `cloudpickle==2.2.0`. I didn't realize that there's only been one release since, but I confirmed that `cloudpickle==2.2.1` does indeed resolve.

All I see in [the changelog](https://github.com/cloudpipe/cloudpickle/blob/master/CHANGES.md) for 2.2.1 is fixing `NamedTuple` pickling; do you think it's possible to provide a more conservative range that would at least include 2.2.1? My guess is some other dependencies pinned `>=2.2.1` because they rely on `NamedTuple` pickling working (but I haven't verified).

Perhaps something like:

```suggestion
cloudpickle~=2.2.0
```

or even

```suggestion
cloudpickle~=2.2
```

depending on what you level of future compatibility you're worried about.

Personally, I think it would also be fine with leaving it uncapped, and users can always bound `cloudpickle` themselves if a breaking release comes out. Preferable to capping it too aggressively, where it requires another release for users to use newer versions. (I quite believe in the rationale in [a popular post on the topic](https://iscinumpy.dev/post/bound-version-constraints/), but in the end you know better how it applies to PyFlink.)

Reply:
Thanks for the explanation. I agree that we should try not to limit the upper bound of python packages. I think it is a relatively safe way to limit `cloudpickle~=2.2.0`. In the future, whenever we upgrade the cloudpickle source package under `lib`, we may need to modify the corresponding python dependencies accordingly.

Reply:
@HuangXingBo Sounds good! I have bound `cloudpickle~=2.2.0` for now, as you suggest. :)